qchem/pennylane_qchem/qchem/obs.py

```diff
@@ -367,9 +367,73 @@ def get_spinZ_matrix_elements(mol_name, hf_data, n_active_electrons=None, n_acti
     return spinz_matrix_elements
 
 
+def get_particle_number_elements(mol_name, hf_data, n_active_electrons=None, n_active_orbitals=None):
+    r"""Reads the Hartree-Fock (HF) electronic structure data file, defines an active
+    space and generates the table required to build the particle number operator
+    :math:`\hat{N}=\sum_\alpha \hat{n}_\alpha`.
+
+    In general, if an active space is defined, the particle number operator reads,
+
+    .. math::
+
+        \hat{N} = 2 n_{\mathrm{docc}} + \sum_\alpha \hat{c}_\alpha^\dagger \hat{c}_\alpha,
+
+    where :math:`2 n_{\mathrm{docc}}` is the number of particles occupying
+    :math:`n_{\mathrm{docc}}` (frozen) core orbitals and :math:`\hat{c}_\alpha^\dagger`
+    (:math:`\hat{c}_\alpha`) is the creation (annihilation) particle operator acting on
+    the :math:`\alpha`-th active spin-orbital.
```

Comment:
Maybe it's good to explain here why the entries of the table are all equal to 1 for this observable

Reply:
I was thinking that maybe for N we don't need the general two-step process of 1) getting the matrix elements (there are actually no matrix elements) and 2) building the observable.

I propose to have a single function returning directly the observable:

```
def particle_number_obs(mol_name, hf_data, n_active_electrons=None, n_active_orbitals=None, mapping=`jordan_wigner`):

    docc_indices, active_indices = structure.active_space(
        mol_name,
        hf_data,
        n_active_electrons=n_active_electrons,
        n_active_orbitals=n_active_orbitals,
    )

    n_spin_orbs = 2 * len(active_indices)

    table = np.ones((n_spin_orbs, 3))
    for alpha in range(n_spin_orbs):
        table[alpha, :2] = alpha
    
    return observable(table, init_term=2 * len(docc_indices), mapping=mapping)
```
I think this makes sense. What do you think?.
Indeed, this may also be a possibility for the other observables.

Reply:
@agran2018 if it makes sense to implement certain observables in a cleaner/more efficient manner, then that is completely how we should be doing it internally 🙂 Things like user experience and UI it is good to have some level of consistency, but that is not required for internal code/implementations if it adds overhead/complexity

Reply:
Thanks @josh146 for the useful recommendations. The problem of the function trying to do too many things was manifesting in different ways as @ixfoduap and @soranjh were also pointing out. I have re-organized things, it is my impression that everything is more consistent now.

The ``observable`` function can be used in most cases internally so we have a single function that returns the observable. It does not make the methodology particular since any user bringing his/her own matrix elements can use ``observable`` independently to build his/her operator. 

I think it's worth to adopt the same strategy for the total spin and Sz. Please, take a look and let me know. Thanks!!.